zds/forum/managers.py

```diff
@@ -148,6 +150,27 @@ def get_all_messages_of_a_user(self, current, target):
 
 
 class TopicReadManager(models.Manager):
+    def _is_read(self, topic, user=None):
+        """
+        Checks if the user has read the **last post** of the topic.
+        Returns false if the user read the topic except its last post.
+        Technically this is done by checking if the user has a `TopicRead` object
+        for the last post of this topic.
+        :param topic: A topic
+        :param user: A user. If undefined, the current user is used.
+        :return:
+        """
+        if user is None:
+            user = get_current_user()
+        if not user.is_authenticated:
+            return False
+        return self.filter(post=topic.last_message, topic=topic, user=user).exists()
+
+    def topic_read(self, topic, user=None):
+        if not hasattr(topic, "_is_read"):
+            setattr(topic, "_is_read", self._is_read(topic, user))
+        return getattr(topic, "_is_read")
```

Comment:
Je me demande si ça ne serait pas plus clair en une seule fonction : 

```suggestion
    def topic_read(self, topic, user=None):
        """
        Checks if the user has read the **last post** of the topic.
        Returns false if the user read the topic except its last post.
        Technically this is done by checking if the user has a `TopicRead` object
        for the last post of this topic.
        :param topic: A topic
        :param user: A user. If undefined, the current user is used.
        :return:
        """

        if hasattr(topic, "_is_read"):
            return getattr(topic, "_is_read")

        if user is None:
            user = get_current_user()
        if not user.is_authenticated:
            return False

        is_read = self.filter(post=topic.last_message, topic=topic, user=user).exists()
        setattr(topic, "_is_read", is_read)
        return is_read
```

Reply:
Il faudrait peut-être aussi trouver un nom plus claire que `topic_read()`. Peut-être `is_topic_last_message_read()` ?

Reply:
fixed